sql: populate pg_type with tables; pg_class.reltypid with type oid

otan opened the pull request:
Resolves #66576

See individual commits for details.

otan committed:
sql: fix vtable index return value for pg_catalog.pg_index

For some reason, populate returned false, nil instead of true, nil
when there was a match for UDTs for the index.

This seemed to still work when I tested it, but just to be safe, I'm
switching the return to true, nil and added a regression test, just in
case...

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/66815)
<!-- Reviewable:end -->

rafiss reviewed (approved):
this lgtm (with one question i had)! i just randomly thought of one concern - is there any risk that someone (or someorm) will look at pg_type now and see these types and try to actually use it as a type? perhaps by casting or something. maybe even if they try, it's not really a problem..

i feel like the risk might not matter, so as long as you agree, lets go ahead with this

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66815#-:-Md04Crl2OYzIJIYSlQf:bnisnpp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/pg_catalog.go, line 550 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66815#-Md0-p731uBAE-yBORvf:-Md0-p731uBAE-yBORvg:bqa1wao) ([raw file](https://github.com/cockroachdb/cockroach/blob/f62f1bd71b494dc1751e387c6b7638f345e51cc0/pkg/sql/pg_catalog.go#L550)):*
> ```
> reloftype
> ```
i think the comment is not correct for `reloftype` -- seems like it's only used for "typed tables" which we don't support. if you feel like fixing it!
___
*[pkg/sql/pg_catalog.go, line 2508 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66815#-Md02k9G-NpcLMLcsDAM:-Md02k9I3SwCJgSurSbe:bv9g5xf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f62f1bd71b494dc1751e387c6b7638f345e51cc0/pkg/sql/pg_catalog.go#L2508)):*
> ```Go
> 					p,
> 					dbContext,
> 					virtualMany,
> ```

is virtualMany what we want here? i'm wondering specifically about privilege checks if the user doesn't permission to view another db/schema or something


<!-- Sent from Reviewable.io -->

otan committed:
sql: populate pg_type with tables; pg_class.reltypid with type oid

It is worth noting we do not populate the array version of the class
(for now) as there is not really a great ID to allocate to it.
PostgreSQL has a similar behaviour for some tables at least, where
typelem and typarray = 0.

Release note (sql change): Populate pg_type table with entries for each
table. Also populate pg_class.reltypid with the corresponding oid in the
pg_type table.

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66815)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[pkg/sql/pg_catalog.go, line 2508 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66815#-Md02k9G-NpcLMLcsDAM:-Md0WjGW3s72L-AwkBsl:byit92w) ([raw file](https://github.com/cockroachdb/cockroach/blob/f62f1bd71b494dc1751e387c6b7638f345e51cc0/pkg/sql/pg_catalog.go#L2508)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

is virtualMany what we want here? i'm wondering specifically about privilege checks if the user doesn't permission to view another db/schema or something
</blockquote></details>

you're right - bad copy pasta.


<!-- Sent from Reviewable.io -->

otan commented:
bors r=rafiss
